core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -163,21 +180,86 @@ private TableMetadata internalApply() {
     this.base = ops.refresh();
 
     Set<Long> idsToRetain = Sets.newHashSet();
-    List<Long> ancestorIds = SnapshotUtil.ancestorIds(base.currentSnapshot(), base::snapshot);
-    if (minNumSnapshots >= ancestorIds.size()) {
+    // Identify refs that should be removed
+    Set<String> refsToRemove = Sets.newHashSet();
+    for (String name : base.refs().keySet()) {
+      SnapshotRef ref = base.ref(name);
+      if (!name.equals(SnapshotRef.MAIN_BRANCH) && ref.maxRefAgeMs() != null) {
+        Snapshot snapshot = base.snapshot(ref.snapshotId());
+        if (removalTimeReference - snapshot.timestampMillis() >= ref.maxRefAgeMs()) {
+          refsToRemove.add(name);
+        }
+
+      }
+
+      if (!refsToRemove.contains(name) && !idsToRemove.contains(ref.snapshotId())) {
```

Comment:
Should `idsToRemove` be allowed to contain snapshots from refs? I think you should be able to remove a snapshot that is a ref and have that drop the ref. I think that should be an error when you expire the snapshot by ID.

Reply:
If I understand what you are saying we are distinguishing the 2 cases, 

1.) the metadata operation to **remove** a snapshot should also drop any refs associated with it. 

2.) **Expiration** of a snapshot which has references should result in failure

This makes sense to me because removing implies a more "blunt" kind of operation where we actually intend on deleting that information. Expiration does not really have that implication, that way unknowingly running an unsafe expiration is prevented.